File: opspilot/frontend/src/api/timeline.ts
import { apiFetch } from './client'

export interface ApiTimelineEvent {
  id: string
  incident_id: string
  type: string
  type_label: string
  title: string
  description: string
  timestamp: string
  relative_time: string
  confidence: number
  is_key_event: boolean
  agent_role: string | null
  metadata: Record<string, string>
}

export interface ApiTimelineResponse {
  incident_id: string
  events: ApiTimelineEvent[]
}

export const timelineApi = {
  get: (incidentId: string): Promise<ApiTimelineResponse> =>
    apiFetch<ApiTimelineResponse>(`/api/timeline/${encodeURIComponent(incidentId)}`),
}

File: opspilot/backend/app/models/timeline.py
"""Pydantic models for the investigation timeline."""
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class TimelineEventType(str, Enum):
    DEPLOYMENT = "deployment"
    INCIDENT = "incident"
    DETECTION = "detection"
    CORRELATION = "correlation"
    ROOT_CAUSE = "root_cause"
    MITIGATION = "mitigation"
    RESOLUTION = "resolution"


class TimelineEvent(BaseModel):
    """A single event on the investigation timeline."""

    id: str
    incident_id: str
    type: TimelineEventType
    type_label: str
    title: str
    description: str
    timestamp: datetime
    relative_time: str = Field(description="Human-readable relative offset, e.g. 'T-20 min'")
    confidence: float = Field(ge=0.0, le=100.0)
    is_key_event: bool = False
    agent_role: str | None = None
    metadata: dict[str, str] = Field(default_factory=dict)


class TimelineResponse(BaseModel):
    incident_id: str
    events: list[TimelineEvent]
